src/gzchunked/mod.rs

```diff
@@ -0,0 +1,210 @@
+
+    /// Retrieves next gzipped block without checking its size.
+    pub fn next_chunk(&mut self) -> std::io::Result<Vec<u8>> {
+        self.encoder.flush()?;
+        let new_encoder = GzEncoder::new(Vec::new(), self.compression.0);
+        let old_encoder = mem::replace(&mut self.encoder, new_encoder);
+        let encoded_data = old_encoder.finish()?;
+
+        Ok(encoded_data)
+    }
+}
+
+impl GzChunkedDecoder {
+    /// Creates a new decoder.
+    pub fn new() -> GzChunkedDecoder {
+        GzChunkedDecoder {
+            queue: VecDeque::new()
+        }
+    }
+
+    /// Decodes next gzchunked block and puts all results into the internal queue.
+    pub fn write(&mut self, buf: &[u8]) -> std::io::Result<()> {
+        let mut decoder = GzDecoder::new(buf);
+        let mut chunked_data_vec: Vec<u8> = Vec::new();
+        decoder.read_to_end(&mut chunked_data_vec)?;
+        let mut chunked_data = chunked_data_vec.as_slice();
+        while !chunked_data.is_empty() {
+            let length = chunked_data.read_u64::<BigEndian>()?;
+            let mut data = vec![0; length as usize];
+            chunked_data.read_exact(data.as_mut_slice())?;
+            self.queue.push_back(data);
+        }
+        Ok(())
+    }
+
+    /// Attempts to retrieve next data piece from queue.
+    /// Returns `None` if the queue is empty.
+    pub fn try_next_data(&mut self) -> Option<Vec<u8>> {
+        self.queue.pop_front()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use rand::{Rng, SeedableRng, rngs::StdRng};
+
+    #[test]
+    fn test_encode_and_decode_empty() {
+        let mut encoder = GzChunkedEncoder::new(GzChunkedCompression::default());
+        let mut decoder = GzChunkedDecoder::new();
+
+        let encoded_block = encoder.next_chunk().unwrap();
+        // This is because the block should at least contain gzip header.
+        assert!(!encoded_block.is_empty());
+        decoder.write(encoded_block.as_slice()).unwrap();
+        decoder.write(encoded_block.as_slice()).unwrap();
+
+        assert!(decoder.try_next_data().is_none());
+    }
+
+    #[test]
+    fn test_encode_and_decode_all_in_one_block() {
+        let mut encoder = GzChunkedEncoder::new(GzChunkedCompression::default());
+        let mut decoder = GzChunkedDecoder::new();
+
+        encoder.write(&[1, 2, 3, 4]).unwrap();
+        encoder.write(&[]).unwrap();
+        encoder.write(&[1, 2, 3, 4, 5, 6, 7, 8]).unwrap();
+        encoder.write(&[1]).unwrap();
+
+        let encoded_block = encoder.next_chunk().unwrap();
+        decoder.write(encoded_block.as_slice()).unwrap();
+
+        assert_eq!([1, 2, 3, 4], decoder.try_next_data().unwrap().as_slice());
+        assert_eq!([] as [u8; 0], decoder.try_next_data().unwrap().as_slice());
+        assert_eq!([1, 2, 3, 4, 5, 6, 7, 8], decoder.try_next_data().unwrap().as_slice());
+        assert_eq!([1], decoder.try_next_data().unwrap().as_slice());
+        assert!(decoder.try_next_data().is_none());
```

Comment:
In Rust (at least in the standard library, which I tend to use as the reference if I am unsure) the expected value is usually specified as the right argument for `assert_eq!`, not the left one.

Also, is there are reason why you do these `.unwrap().as_slice()` rather than comparing with the real value? I would do it like this:

```rust
assert_eq!(decoder.try_next_data(), Some(vec!(1, 2, 3, 4)))`
// (...)
assert_eq!(decoder.try_next_data(), None)
```

I think its both less verbose and clearer what you actually compare against (e.g. explicit `Some`, not an `.unwrap()`). The comparison against `None` with `.is_none()` in general is okay, but if you have a chain of `assert_eq!` before it, I think it just aligns more nicely, if you use explicit comparison against `None`.

Using `vec!` of course causes an extra allocation, but in tests we don't really care and readability should be prioritized.